### DISCUSSION TITLE
Rules returning nil

### DISCUSSION THREAD

**Discussion by inkstak** [palkan/action_policy#270] [Q&A] [answered]
Hi @palkan 

The following issue may be about a bug or about good practice.

**Reproduction Script:** https://gist.github.com/inkstak/448b44bbe97582c26dda00ff043ada4d

```ruby
  def manage?
    if condition_a?
      perform_complicated_check
    elsif condition_b?
      perform_another_complicated_check
    end
  end
```

When none of the conditions match, the rule above will return `nil`.
It's working fine when calling `authorize!` or `allowed_to?`.

However, it fails using the [RSpec DSL](https://actionpolicy.evilmartians.io/#/testing?id=rspec-dsl)

```ruby
  describe_rule :manage? do
    failed "without any conditions"
  end
```
```
  1) UserPolicy manage? without any conditions is denied
     Failure/Error: failed "without any conditions"
     
       Expected to fail but succeed:
       <UserPolicy#manage?: >
```

I thought it was a common practice to accept a predicate method to return `nil`,
but the `failed` method is always expecting `false`.

The workaround is dead simple : always return `false` at least by adding an `else false` to the rule.

But should I add that just to please the specs?
Or should we state that a rule must always return `true` or `false`?

**Comment by palkan** [comment, marked as answer]
> should I add that just to please the specs?

I would say it's not about specs but about the API consistency; predicates must return `true` or `false`, there should be no third boolean value.

Thus, it's the latter:

> should we state that a rule must always return true or false?
